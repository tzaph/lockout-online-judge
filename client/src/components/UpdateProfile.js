import React, { useEffect, useState } from "react";
import { Alert } from "react-bootstrap";
import { useAuth } from "../contexts/AuthContext";
import { getDatabase, ref, get, child } from "firebase/database";
import { Link, useNavigate } from "react-router-dom";
import axios from "axios";

export default function UpdateProfile() {
  const [data, setData] = useState({
    codeforcesHandle: "",
    email: "",
    name: "",
    uid: "",
  });
  const currentUser = useAuth()?.currentUser;
  const updateName = useAuth()?.updateName;
  const updateCodeforcesHandle = useAuth()?.updateCodeforcesHandle;
  const [error, setError] = useState("");
  const [loading, setLoading] = useState(false);
  const [problem, setProblem] = useState({
    contestId: "",
    index: "",
    problemLink: "",
    problemName: "",
  });
  const [submissions, setSubmissions] = useState([]);
  const [codeforcesHandleTaken, setCodeforcesHandleTaken] = useState(false);
  const navigate = useNavigate();

  function listRecentSubmissions(codeforcesHandle) {
    return axios
      .get(
        `https://codeforces.com/api/user.status?handle=${codeforcesHandle}&from=1&count=10`
      )
      .then((res) => {
        setSubmissions(res.data.result);
      })
      .catch((err) => {
        throw err.response.data.comment;
      });
  }

  const isCodeforcesHandleTaken = async (codeforcesHandle) => {
    const db = getDatabase();
    const listOfCodeforcesHandle = [];
    await get(child(ref(db), "users")).then((snapshot) => {
      snapshot.forEach((documentSnapshot) => {
        listOfCodeforcesHandle.push(documentSnapshot.val().codeforcesHandle);
      });
    });
    setCodeforcesHandleTaken(listOfCodeforcesHandle.includes(codeforcesHandle));
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    await updateName(e.target.name.value);
    if (data.codeforcesHandle == e.target.cfHandle.value) {
      setLoading(true);
      setError("");
      navigate("/");
      navigate(0);
      setLoading(false);
      return;
    }
    try {
      await isCodeforcesHandleTaken(e.target.cfHandle.value);
      if (codeforcesHandleTaken) {
        setError(`Codeforces Handle is already taken`);
        return;
      }
      await listRecentSubmissions(e.target.cfHandle.value);
      submissions.map((sub) => {
        if (
          problem.contestId == sub.problem.contestId &&
          problem.problemIndex == sub.problem.index &&
          sub.verdict == "COMPILATION_ERROR"
        ) {
          updateCodeforcesHandle(e.target.cfHandle.value);
          setLoading(true);
          setError("");
          navigate("/");
          navigate(0);
        } else {
          setError(`Failed to verify Codeforces Handle`);
        }
      });
    } catch (err) {
      setError(`Failed to update profile, ${err}`);
    }
    setLoading(false);
  };

  const Getdata = () => {
    get(child(ref(getDatabase()), "users/" + currentUser.uid))
      .then((snapshot) => {
        if (snapshot.exists()) {
          setData(snapshot.val());
        } else {
        }
      })
      .catch((error) => {
        console.error(error);
      });
  };

  const generateProblem = async () => {
    const db = getDatabase();
    await get(child(ref(db), "problems/800")).then((snapshot) => {
      let newProblem =
        snapshot.val().problemsetList[
          Math.floor(Math.random() * snapshot.val().problemsetList.length)
        ];
      let linkUrl =
        "https://codeforces.com/contest/" +
        newProblem.contestId +
        "/problem/" +
        newProblem.index;
      let problemName =
        newProblem.contestId + newProblem.index + " - " + newProblem.name;
      setProblem({
        contestId: newProblem.contestId,
        problemIndex: newProblem.index,
        problemLink: linkUrl,
        problemName: problemName,
      });
    });
  };

  useEffect(() => {
    Getdata();
    generateProblem();
  }, []);

  return (
    <div className="auth-form-container">
      <h2>Update Profile</h2>
      {error && <Alert variant="danger">{error}</Alert>}
      <form className="register-form" onSubmit={handleSubmit}>
        <label for="name">Name</label>
        <input type="name" defaultValue={data.name} id="name" name="name" />
        <label for="cfHandle">Codeforces Handle</label>
        <input
          type="cfHandle"
          defaultValue={data.codeforcesHandle}
          id="cfHandle"
          name="cfHandle"
        />
        <button disabled={loading} type="submit">
          Update
        </button>
      </form>
      Please submit a compile error to{" "}
      <a href={problem.problemLink} target="_blank">
        {problem.problemName}
      </a>{" "}
      to verify your Codeforces handle.
      <div className="link-button">
        <Link to="/">Cancel</Link>
      </div>
    </div>
  );
}
