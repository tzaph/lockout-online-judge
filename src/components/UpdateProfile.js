import React, { useEffect, useState } from 'react'
import { Alert } from 'react-bootstrap';
import { useAuth } from '../contexts/AuthContext';
import { getDatabase, ref, get, child } from "firebase/database";
import { Link, useNavigate } from 'react-router-dom';

export default function UpdateProfile() {
  const [data, setData] = useState({
    codeforcesHandle:"",
    email:"",
    name:"",
    uid:""
  })
  const { currentUser, updateName, updateCodeforcesHandle } = useAuth();
  const [error, setError] = useState('');
  const [loading, setLoading] = useState(false);
  const navigate = useNavigate();
  
  const handleSubmit = async(e) => {
    e.preventDefault();

    try {
      await updateName(name);
      await updateCodeforcesHandle(cfHandle);
      setLoading(true);
      setError('');
      navigate('/');
      navigate(0);
    } catch {
      setError('Failed to update profile');
    }
    setLoading(false);
  }

  const Getdata = () =>{
    get(child(ref(getDatabase()), 'users/' + currentUser.uid)).then((snapshot) => {
      if (snapshot.exists()) {
        setData(snapshot.val())
      } else {
      }
    }).catch((error) => {
      console.error(error);
    });
  }
  useEffect(() => { Getdata(); }, [])

  const [name, setName] = useState(data.name);
  const [cfHandle, setCfhandle] = useState(data.codeforcesHandle);

  return (
    <div className="auth-form-container">
      <h2>Update Profile</h2>
      {error && <Alert variant='danger'>{error}</Alert>}
      <form className="register-form" onSubmit={handleSubmit}>
        <label for="name">Name</label>
        <input
          onChange={(e) => setName(e.target.value)}
          type="name"
          defaultValue={data.name}
          id="name"
          name="name"
        />
        <label for="cfHandle">Codeforces Handle</label>
        <input
          onChange={(e) => setCfhandle(e.target.value)}
          type="cfHandle"
          defaultValue={data.codeforcesHandle}
          id="cfHandle"
          name="cfHandle"
        />
        <button disabled={loading} type="submit">Update</button>
      </form>
      <div className="link-button">
        <Link to="/">Cancel</Link>
      </div>
    </div>
  );
}
