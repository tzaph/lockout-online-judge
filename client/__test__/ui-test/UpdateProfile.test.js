import React from "react";
import renderer from "react-test-renderer";
import UpdateProfile from "../../src/components/UpdateProfile";
import { BrowserRouter as Router } from 'react-router-dom';

describe("<UpdateProfile />", () => {
  it("Renders correctly", () => {
    const tree = renderer.create(<Router><UpdateProfile /></Router>).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
