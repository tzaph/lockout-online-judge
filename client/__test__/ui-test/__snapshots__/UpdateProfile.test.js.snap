// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UpdateProfile /> Renders correctly 1`] = `
<div
  className="auth-form-container"
>
  <h2>
    Update Profile
  </h2>
  <form
    className="register-form"
    onSubmit={[Function]}
  >
    <label
      for="name"
    >
      Name
    </label>
    <input
      defaultValue=""
      id="name"
      name="name"
      type="name"
    />
    <label
      for="cfHandle"
    >
      Codeforces Handle
    </label>
    <input
      defaultValue=""
      id="cfHandle"
      name="cfHandle"
      type="cfHandle"
    />
    <button
      disabled={false}
      type="submit"
    >
      Update
    </button>
  </form>
  <div
    className="link-button"
  >
    <a
      href="/"
      onClick={[Function]}
    >
      Cancel
    </a>
  </div>
</div>
`;
